### PR TITLE
AK: Add static storage size check during move assignments

### PR DESCRIPTION
**Pull request by elvircrn**
Adds static storage size check during move assignments.

### CODE REVIEW

File: AK/Vector.h [hunk 1]
```diff
@@ -41,6 +41,9 @@ struct CanBePlacedInsideVectorHelper<StorageType, false> {
 };
 }
 
+template<size_t inline_capacity, size_t inline_capacity_u>
+concept CanMoveIn = inline_capacity_u <= inline_capacity;
```
Comment (alimpfard) [inline AK/Vector.h:45]: This seems unnecessary, given that...

File: AK/Vector.h [hunk 2]
```diff
@@ -314,15 +321,16 @@ requires(!IsRvalueReference<T>) class Vector {
         MUST(try_prepend(values, count));
     }
 
-    // FIXME: What about assigning from a vector with lower inline capacity?
-    Vector& operator=(Vector&& other)
+    template<class U, size_t inline_capacity_u>
+    requires CanMoveIn<inline_capacity, inline_capacity_u>
```
Comment (alimpfard) [inline AK/Vector.h:325]: ...you can just plop the definition here instead:
```suggestion
    requires(inline_capacity >= inline_capacity_u)
```

File: AK/Vector.h [hunk 3]
```diff
@@ -314,15 +321,16 @@ requires(!IsRvalueReference<T>) class Vector {
...
+    Vector& operator=(Vector<U, inline_capacity_u>&& other)
     {
-        if (this != &other) {
+        if (reinterpret_cast<void*>(this) != reinterpret_cast<void*>(&other)) {
```
Comment (alimpfard) [inline AK/Vector.h:328]: This equality check should only happen if the types are the same, so instead:
```suggestion
        if constexpr (inline_capacity == inline_capacity_u) {
            if (this != &other) {
```

(plus the ensuing indent)

Comment (elvircrn) [reply AK/Vector.h:325]: Should be resolved now.

Comment (alimpfard) [reply AK/Vector.h:325]: Hmm, no, the self-assignment check should only happen when the objects are of the same type (my suggestion was wrong, sorry), it's probably better to check for equality and early return if the types are the same.

Comment (BertalanD) [reply AK/Vector.h:325]: We still need to keep the non-templated `Vector& operator=(Vector&&)` for it to be considered a *move assignment operator* by the language:

https://en.cppreference.com/w/cpp/language/move_assignment
> A move assignment operator is a **non-template** non-static member function with the name `operator=`

We could have a different overload that handles the case with a different inline capacity, see e.g. for `OwnPtr`:
https://github.com/LadybirdBrowser/ladybird/blob/0a3082ef05e8e7a963d3a208acd6efee0ff5c848/AK/OwnPtr.h#L75-L88

**Pull request closed without merging**